clientv3/ordering/kv.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2017 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ordering
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/coreos/etcd/clientv3"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+)
+
+// kvOrdering guarantees that the revision in get responses
+// are equal to or greater than the stored previous known revision.
+type kvOrdering struct {
+	clientv3.KV
+	kvOrderViolation  KvOrderViolationFunc
+	txnOrderViolation TxnOrderViolationFunc
+	prevRev           int64
+}
+
+func NewKV(kv clientv3.KV, kvViolation KvOrderViolationFunc, txnViolation TxnOrderViolationFunc) *kvOrdering {
+	return &kvOrdering{kv, kvViolation, txnViolation, 0}
+}
+
+func (kv *kvOrdering) getPrevRev() int64 {
+	return atomic.LoadInt64(&kv.prevRev)
+}
+
+func (kv *kvOrdering) setPrevRev(currRev int64) {
+	atomic.StoreInt64(&kv.prevRev, currRev)
+}
+
+func (kv *kvOrdering) Get(ctx context.Context, key string, opts ...clientv3.OpOption) (*clientv3.GetResponse, error) {
+	// store prevRev in a local var since concurrent access to kvOrdering could
+	// change the prevRev field by the time the response is received
+	prevRev := kv.getPrevRev()
+	op := clientv3.OpGet(key, opts...)
+	for {
+		r, err := kv.KV.Do(ctx, op)
+		if err != nil {
+			return nil, err
+		}
+		resp := r.Get()
+		if resp.Header.Revision >= prevRev {
+			kv.setPrevRev(resp.Header.Revision)
+			return resp, nil
+		}
+		kv.kvOrderViolation(op, r, prevRev)
+	}
+}
+
+func (kv *kvOrdering) Txn(ctx context.Context) clientv3.Txn {
+	return &txnOrdering{kv.KV.Txn(ctx), kv}
+}
+
+// txnOrdering guarantees that the revision in txn responses
+// are equal to or greater than the stored previous known revision.
+type txnOrdering struct {
+	clientv3.Txn
+	*kvOrdering
+}
+
+func (txn *txnOrdering) Commit() (*clientv3.TxnResponse, error) {
+	// store prevRev in a local var since concurrent access to kvOrdering could
+	// change the prevRev field by the time the response is received
+	prevRev := txn.getPrevRev()
+	for {
+		resp, err := txn.Txn.Commit()
+		if err != nil {
+			return nil, err
+		}
+		var ordered bool = true
+		for _, resOp := range resp.Responses {
```

Comment:
Just for reference, how is the resp.Header.Revision determined? Are the individual responses involved in the process of determining the resp.Header.Revision?